File: perl/p10.pl
#!/usr/bin/perl

$limit = 2000000;
if(scalar($ARGV) > 0){
  $limit = $ARGV[0];
}
$cross_limit = int(sqrt($limit));
@sieve = ();
for($n = 4; $n < $limit; $n += 2){
  $sieve[$n] = 1;
}
for($n = 3; $n <= $cross_limit; $n += 2){
  if ($sieve[$n] == 0){
    for($m = $n * $n; $m < $limit; $m += (2 * $n)){
      $sieve[$m] = 1;
    }
  }
}
$sum = 0;
for($n = 2; $n < $limit; $n++){
  if($sieve[$n] == 0){
    $sum += $n;
  }
}
printf "%llu\n", $sum;

File: ruby/p10.rb
#!/usr/bin/ruby -w

limit = 2000000
if ARGV.length > 0 then
  limit = ARGV[0].to_i
end
cross_limit = Math.sqrt(limit).floor
sieve = Array.new(limit, false)
for n in (4..limit).step(2) do
  sieve[n] = true
end
for n in (3..cross_limit).step(2) do
  if !sieve[n] then
    for m in (n*n..limit).step(2*n) do
      sieve[m] = true
    end
  end
end
sum = 0
for n in (2..limit) do
  if !sieve[n] then
    sum += n
  end
end
puts sum;
